# Assignment 15

# Chapter 11
# 11.5 Writing Mathematical Expressions
## **Question 1. (screenshot: queries and results)**
## Use Production.Product table. We want to get the name, StandardCost and ListPrice for all products that have the word “Touring” and Cost more than “zero”. We want to have a promotional sale for the “Touring” items. Use mathematical expressions on the existing columns to create a new column in the result for the sale price if we offer a 15% discount.


In [1]:
DECLARE @DiscountRate decimal(5,2) = 0.15;

SELECT Name,
       StandardCost,
       ListPrice,
       CAST(ListPrice * (1 - @DiscountRate) AS decimal(10,2)) AS SalePrice
FROM AdventureWorks2022.Production.Product
WHERE Name LIKE '%Touring%'
      AND StandardCost > 0;

(54 rows affected)

Total execution time: 00:00:00.071

Name StandardCost ListPrice SalePrice LL Touring Seat Assembly 98.77 133.34 113.34 ML Touring Seat Assembly 108.99 147.14 125.07 HL Touring Seat Assembly 145.87 196.92 167.38 Touring Front Wheel 96.7964 218.01 185.31 Touring Rear Wheel 108.7844 245.01 208.26 Touring-Panniers, Large 51.5625 125.00 106.25 HL Touring Frame - Yellow, 60 601.7437 1003.91 853.32 LL Touring Frame - Yellow, 62 199.8519 333.42 283.41 HL Touring Frame - Yellow, 46 601.7437 1003.91 853.32 HL Touring Frame - Yellow, 50 601.7437 1003.91 853.32 HL Touring Frame - Yellow, 54 601.7437 1003.91 853.32 HL Touring Frame - Blue, 46 601.7437 1003.91 853.32 HL Touring Frame - Blue, 50 601.7437 1003.91 853.32 HL Touring Frame - Blue, 54 601.7437 1003.91 853.32 HL Touring Frame - Blue, 60 601.7437 1003.91 853.32 LL Touring Frame - Blue, 50 199.8519 333.42 283.41 LL Touring Frame - Blue, 54 199.8519 333.42 283.41 LL Touring Frame - Blue, 58 199.8519 333.42 283.41 LL Touring Frame - Blue, 62 199.8519 333.42 283.41 LL Touring Frame - Yellow, 44 199.8519 333.42 283.41 LL Touring Frame - Yellow, 50 199.8519 333.42 283.41 LL Touring Frame - Yellow, 54 199.8519 333.42 283.41 LL Touring Frame - Yellow, 58 199.8519 333.42 283.41 LL Touring Frame - Blue, 44 199.8519 333.42 283.41 LL Touring Seat/Saddle 12.0413 27.12 23.05 ML Touring Seat/Saddle 17.3782 39.14 33.27 HL Touring Seat/Saddle 23.3722 52.64 44.74 Touring Tire Tube 1.8663 4.99 4.24 Touring Tire 10.8423 28.99 24.64 Touring Pedal 35.9596 80.99 68.84 LL Touring Handlebars 20.464 46.09 39.18 HL Touring Handlebars 40.6571 91.57 77.83 Touring-2000 Blue, 60 755.1508 1214.85 1032.62 Touring-1000 Yellow, 46 1481.9379 2384.07 2026.46 Touring-1000 Yellow, 50 1481.9379 2384.07 2026.46 Touring-1000 Yellow, 54 1481.9379 2384.07 2026.46 Touring-1000 Yellow, 60 1481.9379 2384.07 2026.46 Touring-3000 Blue, 54 461.4448 742.35 631.00 Touring-3000 Blue, 58 461.4448 742.35 631.00 Touring-3000 Blue, 62 461.4448 742.35 631.00 Touring-3000 Yellow, 44 461.4448 742.35 631.00 Touring-3000 Yellow, 50 461.4448 742.35 631.00 Touring-3000 Yellow, 54 461.4448 742.35 631.00 Touring-3000 Yellow, 58 461.4448 742.35 631.00 Touring-3000 Yellow, 62 461.4448 742.35 631.00 Touring-1000 Blue, 46 1481.9379 2384.07 2026.46 Touring-1000 Blue, 50 1481.9379 2384.07 2026.46 Touring-1000 Blue, 54 1481.9379 2384.07 2026.46 Touring-1000 Blue, 60 1481.9379 2384.07 2026.46 Touring-2000 Blue, 46 755.1508 1214.85 1032.62 Touring-2000 Blue, 50 755.1508 1214.85 1032.62 Touring-2000 Blue, 54 755.1508 1214.85 1032.62 Touring-3000 Blue, 44 461.4448 742.35 631.00 Touring-3000 Blue, 50 461.4448 742.35 631.00

## **Question 2. (answer)**

## <u>\- How many Rows?</u>
## 54 Rows

# 11.7 Rounding
## The ROUND function allows you to round a number value to a specific number of decimal places.
## **Question 3. (screenshot: queries and results)**
## We want to report large numbers as rounded to the nearest thousand. Using the Sales.SalesOrderHeader table, get the sales by year rounded to the nearest thousand and formatted as currency. Use the OrderDate field to get your year, and the Subtotal to get your sales. Group and Order by Year. (You should have 4 rows)

In [2]:
SELECT YEAR(OrderDate) AS SalesYear,
       ROUND(SUM(Subtotal) / 1000, 0) * 1000 AS SalesRoundedToThousand
FROM AdventureWorks2022.Sales.SalesOrderHeader
GROUP BY YEAR(OrderDate)
ORDER BY SalesYear;

(4 rows affected)

Total execution time: 00:00:00.025

SalesYear SalesRoundedToThousand 2011 12642000.00 2012 33524000.00 2013 43622000.00 2014 20058000.00

## **Question 4. (answer)**

## \- <u>What is the year with the highest sales value?</u>
## The year 2013 has the highest sales value with $43,622,000.00

## **Question 5. (screenshot: queries and results)**
## We want to report large numbers as rounded to the nearest thousand. Using the Sales.SalesOrderHeader table, get the sales by month of the year with the highest sales value, rounded to the nearest thousand and formatted as currency. Use the OrderDate field to get your month, and the Subtotal to get your sales. 

In [3]:
SELECT MONTH(OrderDate) AS SalesMonth,
       ROUND(SUM(Subtotal), -3) AS RoundedSales
FROM AdventureWorks2022.Sales.SalesOrderHeader
WHERE YEAR(OrderDate) = (SELECT Top 1
                                YEAR(OrderDate)
                         FROM AdventureWorks2022.Sales.SalesOrderHeader
                         GROUP BY YEAR(OrderDate)
                         ORDER BY SUM(Subtotal) DESC
						)
GROUP BY MONTH(OrderDate)
ORDER BY SalesMonth;

(12 rows affected)

Total execution time: 00:00:00.176

SalesMonth RoundedSales 1 2088000.00 2 2317000.00 3 3412000.00 4 2532000.00 5 3246000.00 6 5081000.00 7 4896000.00 8 3334000.00 9 4533000.00 10 4796000.00 11 3312000.00 12 4075000.00

## **Question 6. (answer)**

## <u>\- What is the month with the highest sales value?</u>
## June has the highest sales amount with $5,081,000.00

# 11.8 Rounding Always UP or DOWN (Ceiling, Floor)
## **Question 7. (screenshot: queries and results)**
## The Production.Product table contains the weight of the products that are sold. We want to estimate the shipping weight of products by rounding up to the nearest 10. Get the name, weight and an EstimatedShippingWeight for all the products that have a weight that is not a null value.


In [6]:
SELECT Name,
       Weight,
       CEILING(Weight / 10.0) * 10 AS EstimatedShippingWeight
FROM AdventureWorks2022.Production.Product
WHERE Weight IS NOT NULL
ORDER BY EstimatedShippingWeight DESC;

(205 rows affected)

Total execution time: 00:00:00.015

Name Weight EstimatedShippingWeight LL Road Rear Wheel 1050.00 1050 ML Road Rear Wheel 1000.00 1000 LL Road Front Wheel 900.00 900 HL Road Rear Wheel 890.00 890 ML Road Front Wheel 850.00 850 HL Road Front Wheel 650.00 650 ML Crankset 635.00 640 LL Crankset 600.00 600 HL Crankset 575.00 580 Touring Rim 460.00 460 ML Mountain Rim 450.00 450 LL Road Rim 445.00 450 ML Road Rim 450.00 450 LL Mountain Rim 435.00 440 HL Road Rim 400.00 400 HL Mountain Rim 400.00 400 Rear Brakes 317.00 320 Front Brakes 317.00 320 LL Bottom Bracket 223.00 230 LL Mountain Pedal 218.00 220 ML Mountain Pedal 215.00 220 Rear Derailleur 215.00 220 HL Mountain Pedal 185.00 190 LL Road Pedal 189.00 190 ML Road Pedal 168.00 170 ML Bottom Bracket 168.00 170 HL Bottom Bracket 170.00 170 HL Road Pedal 149.00 150 Front Derailleur 88.00 90 Road-750 Black, 48 20.13 30 Road-750 Black, 52 20.42 30 Touring-2000 Blue, 60 27.90 30 Touring-1000 Yellow, 46 25.13 30 Touring-1000 Yellow, 50 25.42 30 Touring-1000 Yellow, 54 25.68 30 Touring-1000 Yellow, 60 25.90 30 Touring-3000 Blue, 54 29.68 30 Touring-3000 Blue, 58 29.90 30 Touring-3000 Blue, 62 30.00 30 Touring-3000 Yellow, 44 28.77 30 Touring-3000 Yellow, 50 29.13 30 Touring-3000 Yellow, 54 29.42 30 Touring-3000 Yellow, 58 29.79 30 Touring-3000 Yellow, 62 30.00 30 Touring-1000 Blue, 46 25.13 30 Touring-1000 Blue, 50 25.42 30 Touring-1000 Blue, 54 25.68 30 Touring-1000 Blue, 60 25.90 30 Touring-2000 Blue, 46 27.13 30 Touring-2000 Blue, 50 27.42 30 Touring-2000 Blue, 54 27.68 30 Road-750 Black, 58 20.79 30 Touring-3000 Blue, 44 28.77 30 Touring-3000 Blue, 50 29.13 30 Mountain-400-W Silver, 38 26.35 30 Mountain-400-W Silver, 40 26.77 30 Mountain-400-W Silver, 42 27.13 30 Mountain-400-W Silver, 46 27.42 30 Mountain-500 Silver, 40 27.35 30 Mountain-500 Silver, 42 27.77 30 Mountain-500 Silver, 44 28.13 30 Mountain-500 Silver, 48 28.42 30 Mountain-500 Silver, 52 28.68 30 Mountain-500 Black, 40 27.35 30 Mountain-500 Black, 42 27.77 30 Mountain-500 Black, 44 28.13 30 Mountain-500 Black, 48 28.42 30 Mountain-500 Black, 52 28.68 30 Mountain-100 Silver, 38 20.35 30 Mountain-100 Silver, 42 20.77 30 Mountain-100 Silver, 44 21.13 30 Mountain-100 Silver, 48 21.42 30 Mountain-100 Black, 38 20.35 30 Mountain-100 Black, 42 20.77 30 Mountain-100 Black, 44 21.13 30 Mountain-100 Black, 48 21.42 30 Mountain-200 Silver, 38 23.35 30 Mountain-200 Silver, 42 23.77 30 Mountain-200 Silver, 46 24.13 30 Mountain-200 Black, 38 23.35 30 Mountain-200 Black, 42 23.77 30 Mountain-200 Black, 46 24.13 30 Mountain-300 Black, 38 25.35 30 Mountain-300 Black, 40 25.77 30 Mountain-300 Black, 44 26.13 30 Mountain-300 Black, 48 26.42 30 Road-250 Red, 44 14.77 20 Road-250 Red, 48 15.13 20 Road-250 Red, 52 15.42 20 Road-250 Red, 58 15.79 20 Road-250 Black, 44 14.77 20 Road-250 Black, 48 15.13 20 Road-250 Black, 52 15.42 20 Road-250 Black, 58 15.68 20 Road-550-W Yellow, 38 17.35 20 Road-550-W Yellow, 40 17.77 20 Road-550-W Yellow, 42 18.13 20 Road-550-W Yellow, 44 18.42 20 Road-550-W Yellow, 48 18.68 20 Road-150 Red, 62 15.00 20 Road-150 Red, 44 13.77 20 Road-150 Red, 48 14.13 20 Road-150 Red, 52 14.42 20 Road-150 Red, 56 14.68 20 Road-450 Red, 58 17.79 20 Road-450 Red, 60 17.90 20 Road-450 Red, 44 16.77 20 Road-450 Red, 48 17.13 20 Road-450 Red, 52 17.42 20 Road-650 Red, 58 19.79 20 Road-650 Red, 60 19.90 20 Road-650 Red, 62 20.00 20 Road-650 Red, 44 18.77 20 Road-650 Red, 48 19.13 20 Road-650 Red, 52 19.42 20 Road-650 Black, 58 19.79 20 Road-650 Black, 60 19.90 20 Road-650 Black, 62 20.00 20 Road-650 Black, 44 18.77 20 Road-650 Black, 48 19.13 20 Road-650 Black, 52 19.42 20 Road-350-W Yellow, 40 15.35 20 Road-350-W Yellow, 42 15.77 20 Road-350-W Yellow, 44 16.13 20 Road-350-W Yellow, 48 16.42 20 Road-750 Black, 44 19.77 20 ML Mountain Frame - Black, 40 2.77 10 ML Mountain Frame - Black, 44 2.81 10 ML Mountain Frame - Black, 48 2.85 10 ML Road Frame-W - Yellow, 40 2.22 10 ML Road Frame-W - Yellow, 42 2.26 10 ML Road Frame-W - Yellow, 44 2.30 10 ML Road Frame-W - Yell

## **Question 8. (answer)**
## <u>- Which product weighs the most?</u>
## The LL Road Rear Wheel is the product with the most wieght with 1050

# 11.11 Treat NULLS as Zeros (if you are sure you want to!)
## **Question 9. (screenshot: queries and results)**
## SELECT SpecialOfferID, MaxQty, COALESCE(MaxQty, 0) AS MaxQtyAlt FROM Sales.SpecialOffer


In [7]:
SELECT SpecialOfferID, MaxQty, COALESCE(MaxQty, 0) AS MaxQtyAlt 
FROM AdventureWorks2022.Sales.SpecialOffer

(16 rows affected)

Total execution time: 00:00:00.013

SpecialOfferID MaxQty MaxQtyAlt 1 NULL 0 2 14 14 3 24 24 4 40 40 5 60 60 6 NULL 0 7 NULL 0 8 NULL 0 9 NULL 0 10 NULL 0 11 NULL 0 12 NULL 0 13 NULL 0 14 NULL 0 15 NULL 0 16 NULL 0

## **Question 10. (explanation)**
## <u>What are some reasons that we should be careful about converting NULL values to ZEROs?</u>
## Converting NULLs to zeros might mess with data integrity, especially if those NULLs were meant to show something specific (like "no data" instead of "zero"). Treating missing data as zeros could skew analysis or reports, giving a misleading picture of the situation.